pkg/kubelet/volumemanager/reconciler/reconciler_test.go

```diff
@@ -1221,11 +1223,12 @@ func Test_UncertainDeviceGlobalMounts(t *testing.T) {
 			supportRemount:         true,
 		},
 	}
-
 	for _, mode := range []v1.PersistentVolumeMode{v1.PersistentVolumeBlock, v1.PersistentVolumeFilesystem} {
 		for _, tc := range tests {
 			testName := fmt.Sprintf("%s [%s]", tc.name, mode)
-			t.Run(testName+"[", func(t *testing.T) {
+			wg.Add(1)
+			go t.Run(testName+"[", func(t *testing.T) {
```

Comment:
don't start `t.Run` with a goroutine, call `t.Parallel()` at the top of the inner test function to tell go to run the tests in parallel (and it handles waiting until all are done to exit, no need for a waitgroup).

you'll also need to switch `mode` and `tc` vars to be local to this loop or the closure will look at the wrong struct

```go
modes := []v1.PersistentVolumeMode{v1.PersistentVolumeBlock, v1.PersistentVolumeFilesystem}
for modeIndex := range modes {
  for tcIndex := range tests {
    mode := modes[modeIndex]
    tc := tests[tcIndex]
    testName := fmt.Sprintf("%s [%s]", tc.name, mode)
    t.Run(testName+"[", func(t *testing.T) {
      t.Parallel()
...
```

Reply:
same comments apply below

Reply:
I'll defer to storage reviewers to be sure these are safe to run in parallel (do they share/modify any global state / test dirs?)

Reply:
+1 to using `t.Parallel` and saving `tc` and `mode` variables. These tests should be okay to run as parallel tests. I don't think they share any state and I tested the changes with `go test -race`.

Reply:
Also waiting for specific state from outside is bit tricky - these tests are trying to get reconciler to first timeout a mount and then fail a mount, so these operations should happen in that order. We can't easily wait for these operations to happen in that order from outside (because time we try and inspect for intermediate state, we might already have missed it) , we could wait for end state but then in some cases end state might be indistinguishable from beginning state (say volume should be marked as unmounted, if second mount failed but first mount timed out - but volume was unmounted to begin with!).

Reply:
is `kubeletPodsDir = "fake-dir"` shared between reconcilers? when I stressed the tests I saw errors about `fake-dir` not existing, but had a hard time telling if that was 1) an existing flake/race, 2) the parallel instances stomping on each other, or 3) expected errors because we're using a fake dir

Reply:
When I used `t.Parallel()` to run the test concurrently, I found that it was not stable. Here is the error log:
```
--- FAIL: Test_UncertainVolumeMountState (0.00s)
    --- FAIL: Test_UncertainVolumeMountState/failed_operation_should_result_in_not-mounted_volume_[Block][ (2.02s)
        reconciler_test.go:1552: Error verifying UnMountDeviceCallCount: No Unmapper have expected TearDownDeviceCallCount. Expected: <1>.
I0203 11:04:43.183588    9998 reconciler.go:224] operationExecutor.VerifyControllerAttachedVolume started for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") 
I0203 11:04:43.183609    9998 reconciler.go:157] Reconciler: start to sync state
E0203 11:04:43.183638    9998 reconciler.go:389] Cannot get volumes from disk open fake-dir: no such file or directory
I0203 11:04:43.183715    9998 operation_generator.go:1357] Controller attach succeeded for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") device path: "/fake/path"
I0203 11:04:43.189019    9998 operation_generator.go:556] MountVolume.WaitForAttach entering for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") DevicePath "/fake/path"
I0203 11:04:43.189113    9998 operation_generator.go:565] MountVolume.WaitForAttach succeeded for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") DevicePath "/dev/sdb"
I0203 11:04:43.189133    9998 operation_generator.go:594] MountVolume.MountDevice succeeded for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") device mount path ""
I0203 11:04:43.197783    9998 operation_generator.go:556] MountVolume.WaitForAttach entering for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") DevicePath "/fake/path"
I0203 11:04:43.336732    9998 operation_generator.go:565] MountVolume.WaitForAttach succeeded for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") DevicePath "/dev/sdb"
I0203 11:04:43.336865    9998 reconciler.go:196] operationExecutor.UnmountVolume started for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1uid" (UID: "pod1uid") 
I0203 11:04:43.359599    9998 operation_generator.go:797] UnmountVolume.TearDown succeeded for volume "fake-plugin/fake-device1" (OuterVolumeSpecName: "volume-name") pod "pod1uid" (UID: "pod1uid"). InnerVolumeSpecName "volume-name". PluginName "fake-plugin", VolumeGidValue ""
I0203 11:04:43.402408    9998 reconciler.go:196] operationExecutor.UnmountVolume started for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1uid" (UID: "pod1uid") 
I0203 11:04:44.142509    9998 operation_generator.go:797] UnmountVolume.TearDown succeeded for volume "fake-plugin/fake-device1" (OuterVolumeSpecName: "volume-name") pod "pod1uid" (UID: "pod1uid"). InnerVolumeSpecName "volume-name". PluginName "fake-plugin", VolumeGidValue ""
I0203 11:04:44.451165    9998 reconciler.go:312] operationExecutor.UnmountDevice started for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") on node "mynodename" 
I0203 11:04:44.492100    9998 reconciler_test.go:1806] UnmountDevice called
I0203 11:04:44.572667    9998 operation_generator.go:891] UnmountDevice succeeded for volume "volume-name" %!(EXTRA string=UnmountDevice succeeded for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") on node "mynodename" )
I0203 11:04:44.947129    9998 operation_generator.go:556] MountVolume.WaitForAttach entering for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") DevicePath "/dev/sdb"
I0203 11:04:44.947159    9998 operation_generator.go:565] MountVolume.WaitForAttach succeeded for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") DevicePath "/dev/sdb"
I0203 11:04:44.947181    9998 operation_generator.go:594] MountVolume.MountDevice succeeded for volume "volume-name" (UniqueName: "fake-plugin/fake-device1") pod "pod1" (UID: "pod1uid") device mount path ""
FAIL
FAIL    k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler  31.673s
FAIL
```
 We specify `kubeletPodsDir` for subtests separately?

Reply:
The error:

```
E0203 11:04:43.183638    9998 reconciler.go:389] Cannot get volumes from disk open fake-dir: no such file or directory
```

is expected - this happens because the pod volume directory does not actually get created until `Setup` function is called and that code tries to "restore" kubelet state by reading volumes from the disk (but pod-dir does not exist yet). But @liggitt is right - sharing `kubeletPodsDir` is problematic, specially because pod UUIDs are not unique at all. 

I ended up using following code:

```
uniqueTestString := fmt.Sprintf("local-mount-%s", testName)
uniquePodDir := fmt.Sprintf("%s-%x", kubeletPodsDir, md5.Sum([]byte(uniqueTestString)))
```

I still saw some failures however when I ran with:

```
stress -p 12 ./reconciler.test -test.run "Test_UncertainVolumeMountState"
```

but when I run with:

```
while do; go test -count=1 -race -run Test_UncertainVolumeMountState\$ ./pkg/kubelet/volumemanager/reconciler -v; sleep 1; done
```

For 1 hour or so - nothing failed.

Reply:
stress runs in parallel and puts the system under more load... that tends to expose flaky tests

Reply:
> When I used t.Parallel() to run the test concurrently, I found that it was not stable.

That's not a valid reason to use manual go routines, but to fix the test instead.

Reply:
@gnufied I run the stress test too and these are my results:

```
stress -p 12 ./reconciler.test -test.run Test_UncertainVolumeMountState
// ...
1h0m0s: 684 runs so far, 2 failures (0.29%)
```

The error logs (1/2):

```
/tmp/go-stress-20210208T222813-811969047
I0208 23:18:21.965745 2762600 reconciler.go:224] operationExecutor.VerifyControllerAttachedVolume started for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid")
I0208 23:18:21.965934 2762600 reconciler.go:157] Reconciler: start to sync state
E0208 23:18:21.965990 2762600 reconciler.go:389] Cannot get volumes from disk open fake-dir: no such file or directory
I0208 23:18:21.971687 2762600 operation_generator.go:1357] Controller attach succeeded for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid") device path: "fake/path"
I0208 23:18:21.972429 2762600 operation_generator.go:973] MapVolume.WaitForAttach entering for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid") DevicePath "fake/path"
I0208 23:18:21.972701 2762600 operation_generator.go:982] MapVolume.WaitForAttach succeeded for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid") DevicePath "/dev/sdb"
E0208 23:18:21.979871 2762600 nestedpendingoperations.go:301] Operation for "{volumeName:fake-plugin/timeout-setup-volume podName: nodeName:}" failed. No retries permitted until 2021-02-08 23:18:22.479637129 +0000 UTC m=+0.819160815 (durationBeforeRetry 500ms). Error: "MapVolume.MapPodDevice failed for volume \"timeout-setup-volume\" (UniqueName: \"fake-plugin/timeout-setup-volume\") pod \"pod1\" (UID: \"pod1uid\") : time out on setup"
I0208 23:18:22.074454 2762600 operation_generator.go:1170] UnmapVolume succeeded for volume "fake-plugin/timeout-setup-volume" (OuterVolumeSpecName: "timeout-setup-volume") pod "pod1uid" (UID: "pod1uid"). InnerVolumeSpecName "timeout-setup-volume". PluginName "fake-plugin", VolumeGidValue ""
I0208 23:18:22.075022 2762600 reconciler.go:196] operationExecutor.UnmountVolume started for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1uid" (UID: "pod1uid")
I0208 23:18:22.07538
```

(2/2):

```
/tmp/go-stress-20210208T222813-396723850
I0208 23:20:27.869411 2764274 reconciler.go:224] operationExecutor.VerifyControllerAttachedVolume started for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid")
I0208 23:20:27.869586 2764274 reconciler.go:157] Reconciler: start to sync state
E0208 23:20:27.869641 2764274 reconciler.go:389] Cannot get volumes from disk open fake-dir: no such file or directory
I0208 23:20:27.869971 2764274 operation_generator.go:1357] Controller attach succeeded for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid") device path: "fake/path"
I0208 23:20:27.870633 2764274 operation_generator.go:973] MapVolume.WaitForAttach entering for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid") DevicePath "fake/path"
I0208 23:20:27.870750 2764274 operation_generator.go:982] MapVolume.WaitForAttach succeeded for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1" (UID: "pod1uid") DevicePath "/dev/sdb"
E0208 23:20:27.871156 2764274 nestedpendingoperations.go:301] Operation for "{volumeName:fake-plugin/timeout-setup-volume podName: nodeName:}" failed. No retries permitted until 2021-02-08 23:20:28.370961228 +0000 UTC m=+0.978662642 (durationBeforeRetry 500ms). Error: "MapVolume.MapPodDevice failed for volume \"timeout-setup-volume\" (UniqueName: \"fake-plugin/timeout-setup-volume\") pod \"pod1\" (UID: \"pod1uid\") : time out on setup"
I0208 23:20:27.972156 2764274 reconciler.go:196] operationExecutor.UnmountVolume started for volume "timeout-setup-volume" (UniqueName: "fake-plugin/timeout-setup-volume") pod "pod1uid" (UID: "pod1uid")
I0208 23:20:27.972268 2764274 operation_generator.go:1170] UnmapVolume succeeded for volume "fake-plugin/timeout-setup-volume" (OuterVolumeSpecName: "timeout-setup-volume") pod "pod1uid" (UID: "pod1uid"). InnerVolumeSpecName "timeout-setup-volume". PluginName "fake-plugin", VolumeGidValue ""
I0208 23:20:27.97373
```

I didn't add the `uniqueTestString` or `uniquePodDir` variables but I added the parallel code @liggitt suggested

Reply:
I made a mistake and tested the wrong test function, I added `t.Parallel()` and the captured the variables inside the closure to `Test_UncertainDeviceGlobalMounts` instead of `Test_UncertainVolumeMountState`, please disregard the previous stress test results

Reply:
I fixed these tests here - https://github.com/kubernetes/kubernetes/pull/99174/files